### PR TITLE
CosmosDB: Initial implementation to build and use SAS tokens

### PR DESCRIPTION
**Pull request by milismsft**
Initial implementation that allows for building and using SAS like tokens that were generated on the client side using primary or secondary Cosmos DB keys. These are in some ways similar to current resource tokens but without the overhead of having them being generated on the Cosmos service side.

The "contract" for the new authorization header is:
`
"type=sas&ver=1.0&sig=BASE64_HMAC_ENCODED_PAYLOAD_USING_COSMOS_KEY;BASE64_PAYLOAD_VALUE"
`

Where payload value is the Base64 URL encoding of a string containing all the following values:
user\n
userTag\n
resourcePathPrefix\n
BASE64_ENCODED_PartitionKeyValues_CommaSeparated\n
epochStartTime\n
epochExpiryTime\n
keyKind\n
shortAsHexControlPlaneReaderScope\n
shortAsHexControlPlaneWriterScope\n
intAsHexDataPlaneReaderScope\n
intAsHexDataPlaneWriterScope\n

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/sastokens/SasTokenImpl.java [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+
+        if (this.dataPlaneReaderScope == 0 && this.dataPlaneWriterScope == 0) {
+            this.dataPlaneReaderScope |= SCOPE_CONTAINER_READ_ALL_ACCESS.value();
+        }
+
+        StringBuilder payload = new StringBuilder(this.user).append("\n")
+            .append(this.userTag).append("\n")
+            .append(this.resourcePath).append("\n")
+            .append(partitionRanges).append("\n")
+            .append(String.format("%X", this.startTime.getEpochSecond())).append("\n")
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/sastokens/SasTokenImpl.java:178]: generally `String.format(.)` will do string regex which is computationally expensive. hence we should avoid in the per request hot path.

Comment (milismsft) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/sastokens/SasTokenImpl.java:178]: This is only used as part of constructing the token's payload which is one time thing.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/sastokens/SasTokenAuthorizationHelper.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+public class SasTokenAuthorizationHelper {
+    private static final Logger logger = LoggerFactory.getLogger(SasTokenAuthorizationHelper.class);
+
+    /**
+     * This method help to differentiate between master key and SAS token.
+     *
+     * @param token SAS token provided.
+     * @return Whether given token is a SAS token or not.
+     */
+    public static boolean isSasToken(String token) {
```
[the remark this answers is not in the document]
Comment (milismsft) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/sastokens/SasTokenAuthorizationHelper.java:23]: Correct; similar to ResourceTokenAuthorizationHelper.isResourceToken(), only used during the construction of the CosmosClient's instance.

**Comment by JonathanGiles** [comment]
@milismsft I'm curious if this PR is still relevant? Thanks!

**Comment by milismsft** [comment]
@JonathanGiles yes, this is still an ongoing work item. We are currently working on finishing the Cosmos Service side implementation, once that is in, I will update and merge this PR.

**Pull request closed without merging**